chia/types/blockchain_format/coin.py

```diff
@@ -53,11 +53,11 @@ def __bytes__(self) -> bytes:  # pylint: disable=E0308
         assert False
 
 
-def hash_coin_list(coin_list: List[Coin]) -> bytes32:
-    coin_list.sort(key=lambda x: x.name_str, reverse=True)
```

Comment:
this is the main change in this PR. note how this list sorts `Coins`, and each time the sorting algorithm is comparing two coins, we compute the coin_id (which is not trivial) and then turn it into a *hex string*, to compare.

This patch converts the list of coins to a list of `bytes32` objects instead, and we sort them directly. No need to recompute them every time we need to compare them. There's also no point in generating a string, and comparing strings.

Reply:
this new version of this function does not convert the name to a hex string before sorting. In order to ensure that there's no unexpected reason the hex string would sort different than the binary name, there's a test here: https://github.com/Chia-Network/chia-blockchain/pull/11421/files#diff-9e2f2e07ef5b18a6367ed8a05b8c369930af75267ce08c0229d0ed6da9bad7acR27

Reply:
https://docs.python.org/3/library/stdtypes.html#list.sort
> The key corresponding to each item in the list is calculated once and then used for the entire sorting process.

But in general I'm all for using the proper form of the data (bytes) instead of a serialization or human representation form (hexadecimal string).

Reply:
interesting. I would expect the common case is that you pick one field in your type in the lambda. But that suggests that the sort function still copies that field just to avoid invoking the lambda multiple times.